### PR TITLE
[SPARK-25704][CORE] Allocate a bit less than Int.MaxValue

### PR DESCRIPTION
**Pull request by squito**
JVMs don't you allocate arrays of length exactly Int.MaxValue, so leave
a little extra room.  This is necessary when reading blocks >2GB off
the network (for remote reads or for cache replication).

Unit tests via jenkins, ran a test with blocks over 2gb on a cluster

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 1]
```diff
@@ -195,7 +196,11 @@ object ChunkedByteBuffer {
     val is = new FileInputStream(file)
     ByteStreams.skipFully(is, offset)
     val in = new LimitedInputStream(is, length)
-    val chunkSize = math.min(maxChunkSize, length).toInt
+    // Though in theory you should be able to index into an array of size Int.MaxValue, in practice
+    // jvms don't let you go up to limit.  It seems you may only need - 2, but we leave a little
+    // extra room.
+    val maxArraySize = Int.MaxValue - 512
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: Looks like the max is safely more like `Int.MaxValue - 5` from things like https://www.quora.com/What-is-the-maximum-size-of-the-array-in-Java You could probably push it further, but whatever.

Comment (kiszk) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: We already had seen the similar problem in other places. Can we use [this value](https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java#L52) here, too?

Comment (squito) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: great suggestion, thanks, just updated

**Comment by squito** [comment]
Jenkins, retest this please

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
LGTM

**Comment by squito** [comment]
merged to master / 2.4.

Something went wrong w/ the merge script when putting it on 2.4 (I think I accidentally had a dirty working dir at the time), but it was actually a clean cherry pick, so I pushed it manually here:
https://github.com/apache/spark/commit/1001d2314275c902da519725da266a23b537e33a

**Pull request closed without merging**